File: clients/cabinet/shared/auth/oauth.ts
// Google OAuth2 authorize-request construction: PKCE (S256), state, and nonce,
// minted at the BFF with the Web Crypto API (Node runtime). The code verifier
// never leaves the server (stored with the OAuth transaction); only the derived
// challenge goes to Google.

const AUTHORIZE_ENDPOINT = "https://accounts.google.com/o/oauth2/v2/auth";
const SCOPE = "openid email profile";

export interface OAuthChallenge {
  state: string;
  nonce: string;
  codeVerifier: string;
  codeChallenge: string;
}

function base64url(buf: ArrayBuffer | Uint8Array): string {
  const bytes = buf instanceof Uint8Array ? buf : new Uint8Array(buf);
  return Buffer.from(bytes).toString("base64url");
}

function randomToken(bytes = 32): string {
  const buf = new Uint8Array(bytes);
  crypto.getRandomValues(buf);
  return base64url(buf);
}

async function sha256(input: string): Promise<string> {
  const digest = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(input));
  return base64url(digest);
}

/** Mint a fresh PKCE verifier/challenge plus an anti-forgery state and nonce. */
export async function newChallenge(): Promise<OAuthChallenge> {
  const codeVerifier = randomToken(32);
  return {
    state: randomToken(16),
    nonce: randomToken(16),
    codeVerifier,
    codeChallenge: await sha256(codeVerifier),
  };
}

/** Build the Google authorize URL to redirect the browser to. */
export function authorizeUrl(opts: { clientId: string; redirectUri: string; state: string; nonce: string; codeChallenge: string }): string {
  const params = new URLSearchParams({
    client_id: opts.clientId,
    redirect_uri: opts.redirectUri,
    response_type: "code",
    scope: SCOPE,
    state: opts.state,
    nonce: opts.nonce,
    code_challenge: opts.codeChallenge,
    code_challenge_method: "S256",
    access_type: "online",
    prompt: "select_account",
  });
  return `${AUTHORIZE_ENDPOINT}?${params.toString()}`;
}

/** Keep a post-login redirect target same-origin to defeat open-redirects. */
export function safeReturnTo(raw: string | null): string {
  if (!raw || !raw.startsWith("/")) return "/";
  // Reject protocol-relative ("//evil", "/\evil") and any backslash, which some
  // browsers normalize to "/" — both can escape to another origin.
  if (raw[1] === "/" || raw[1] === "\\" || raw.includes("\\")) return "/";
  return raw;
}

File: clients/cabinet/backend/src/oauth.rs
use std::collections::HashMap;

use sha2::{Digest, Sha256};
use tokio::sync::Mutex;

use crate::util::{base64url, now_secs, random_token};

const AUTHORIZE_ENDPOINT: &str = "https://accounts.google.com/o/oauth2/v2/auth";
const SCOPE: &str = "openid email profile";
/// The OAuth handshake (PKCE/state/nonce) lives at most this long between authorize and callback.
pub const OAUTH_TX_TTL: i64 = 600;

/// A fresh PKCE verifier/challenge plus anti-forgery state and nonce.
pub struct Challenge {
	pub state: String,
	pub nonce: String,
	pub code_verifier: String,
	pub code_challenge: String,
}

impl Challenge {
	pub fn new() -> Self {
		let code_verifier = random_token(32);
		let code_challenge = base64url(&Sha256::digest(code_verifier.as_bytes()));
		Self {
			state: random_token(16),
			nonce: random_token(16),
			code_verifier,
			code_challenge,
		}
	}
}

/// Build the Google authorize URL to redirect the browser to.
pub fn authorize_url(client_id: &str, redirect_uri: &str, state: &str, nonce: &str, code_challenge: &str) -> String {
	let query = form_urlencoded::Serializer::new(String::new())
		.append_pair("client_id", client_id)
		.append_pair("redirect_uri", redirect_uri)
		.append_pair("response_type", "code")
		.append_pair("scope", SCOPE)
		.append_pair("state", state)
		.append_pair("nonce", nonce)
		.append_pair("code_challenge", code_challenge)
		.append_pair("code_challenge_method", "S256")
		.append_pair("access_type", "online")
		.append_pair("prompt", "select_account")
		.finish();
	format!("{AUTHORIZE_ENDPOINT}?{query}")
}

/// Keep a post-login redirect target same-origin to defeat open-redirects (port of `safeReturnTo`).
pub fn safe_return_to(raw: Option<&str>) -> String {
	let Some(raw) = raw else { return "/".to_string() };
	if !raw.starts_with('/') {
		return "/".to_string();
	}
	// Reject protocol-relative ("//evil", "/\evil") and any backslash.
	let second = raw.as_bytes().get(1).copied();
	if second == Some(b'/') || second == Some(b'\\') || raw.contains('\\') {
		return "/".to_string();
	}
	raw.to_string()
}

/// One in-flight OAuth login transaction, bound to the `ev_oauth_tx` cookie.
#[derive(Clone)]
pub struct OAuthTx {
	pub state: String,
	pub nonce: String,
	pub code_verifier: String,
	pub return_to: String,
	created_at: i64,
}

/// The OAuth transaction store. In-process map (single-instance/dev), keyed by the
/// HttpOnly `ev_oauth_tx` cookie so only the browser that started the flow can complete it.
pub struct OAuthTxStore {
	txns: Mutex<HashMap<String, OAuthTx>>,
}

impl OAuthTxStore {
	pub fn new() -> Self {
		Self { txns: Mutex::new(HashMap::new()) }
	}

	/// Store a transaction, returning its id (the `ev_oauth_tx` cookie value).
	pub async fn put(&self, state: String, nonce: String, code_verifier: String, return_to: String) -> String {
		let id = random_token(32);
		let tx = OAuthTx {
			state,
			nonce,
			code_verifier,
			return_to,
			created_at: now_secs(),
		};
		self.txns.lock().await.insert(id.clone(), tx);
		id
	}

	/// Read + consume the transaction for `id`, if present and unexpired.
	pub async fn take(&self, id: &str) -> Option<OAuthTx> {
		let tx = self.txns.lock().await.remove(id)?;
		(now_secs() - tx.created_at <= OAUTH_TX_TTL).then_some(tx)
	}
}
